### PR TITLE
Plan clone refactor

### LINKED ISSUE #1183: Changing cache logic causes test failures
Currently we run queries on the cloned ExecutionPlan, and store the original in the cache. Swapping these should cause identical behavior, but instead several flow tests fail if this is done.

### LINKED ISSUE #1270: ExecutionPlan_Clone not producing identical plans
Issue #1268 displays a query which causes a crash in the `ExecutionPlan_Clone` logic:
```
  MATCH (u:User)-->(g:Group)-->(p:Parcel)
  WHERE u.uuid = ${ bearer.uuid }
  AND g.uuid = ${ uuid } AND p.uuid = ${ parcel_id }
  WITH u, g, p
  OPTIONAL MATCH (g)<--(a:Assign)
  WITH DISTINCT u, g, p, a
  MATCH (tenant:User)-[rel:IS_TENANT]->(p)
  WHERE tenant.uuid = ${ tenant_id }
  AND ( (u)-->(a)-->(p) OR (u)-[:IS_OWNER]->(g) )
  RETURN DISTINCT rel AS is_in, tenant AS the_tenant, p AS parcel
```
This crash occurs at https://github.com/RedisGraph/RedisGraph/blob/master/src/execution_plan/execution_plan_clone.c#L55 , because `current_segment->root` is NULL.

Many variations of this query cause crashes or inappropriate ExecutionPlans to be produced by the clone routine. For example, on master, we see the invalid ExecutionPlan:
```
GRAPH.EXPLAIN "MATCH (u:User)-->(g:Group)-->(p:Parcel) WHERE u.uuid = 5 AND g.uuid = 5 AND p.uuid = 32 WITH u, g, p MATCH (tenant:User)-[rel:IS_TENANT]->(p) WHERE tenant.uuid = 2 AND (u)-->(p) RETURN DISTINCT *"
1) "Results"
2) "    Distinct"
3) "        Project"
4) "            Filter"
5) "                Conditional Traverse | (p)<-[rel:IS_TENANT]-(tenant:User)"
6) "                    Project"
7) "                        Filter"
```
On commit 98043a517 (the last prior to cache introduction), the same query produces the plan:
```
 1) "Results"
 2) "    Distinct"
 3) "        Project"
 4) "            Filter"
 5) "                Conditional Traverse | (p)<-[rel:IS_TENANT]-(tenant:User)"
 6) "                    Project"
 7) "                        Semi Apply"
 8) "                            Conditional Traverse | (g:Group)->(p:Parcel)"
 9) "                                Conditional Traverse | (u:User)->(g:Group)"
10) "                                    Node By Label Scan | (u:User)"
11) "                            Expand Into | (u)->(p)"
12) "                                Argument"
```

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Updates ExecutionPlan Clone and Free logic to crawl the op tree rather than storing ExecutionPlan segments. Removes extraneous members from the ExecutionPlan struct.

Resolves #1270, does not resolve #1183, which should be looked into further!

### CODE REVIEW

File: src/execution_plan/execution_plan_clone.c [hunk 1]
```diff
@@ -8,92 +8,69 @@
...
-/* This function clones execution plan by cloning each segment in the execution plan as a unit.
- * Each segment has its own filter tree, record mapping, query graphs and ast segment, that compose
- * a single logical execution unit, together with the segment operations.
- * The ast segment is shallow copied while all the other objects are deep cloned.
- */
+/* This function clones the input ExecutionPlan by recursively visiting its tree of ops.
+ * When an op is encountered that was constructed as part of a different ExecutionPlan segment, that segment
+ * and its internal members (FilterTree, record mapping, query graphs, and AST segment) are also cloned. */
 ExecutionPlan *ExecutionPlan_Clone(const ExecutionPlan *template) {
 	if(template == NULL) return NULL;
```
Comment (swilly22) [inline src/execution_plan/execution_plan_clone.c:66]: `ASSERT(template != NULL)` ?

File: src/execution_plan/execution_plan_clone.c [hunk 2]
```diff
@@ -8,92 +8,69 @@
...
+	// Store the original AST pointer.
+	AST *master_ast = QueryCtx_GetAST();
 	// Verify that the execution plan template is not prepared yet.
 	assert(template->prepared == false && "Execution plan cloning should be only on templates");
```
Comment (swilly22) [inline src/execution_plan/execution_plan_clone.c:70]: Use debug assert, `ASSERT`

File: src/execution_plan/execution_plan_clone.c [hunk 3]
```diff
@@ -8,92 +8,69 @@
...
+
+	return clone_current;
+}
+
+static ExecutionPlan *_ExecutionPlan_Clone(const ExecutionPlan *template) {
+	OpBase *clone_root = _CloneOpTree(NULL, template->root, NULL);
+	// The "master" execution plan is the one constructed with the root op.
+	ExecutionPlan *clone = (ExecutionPlan *)clone_root->plan;
+	// Set the root op pointer.
+	clone->root = clone_root;
```
Comment (swilly22) [inline src/execution_plan/execution_plan_clone.c:57]: isn't `clone_root->plan->root` is `clone_root` ?
in which case we can return `clone` ?

File: src/execution_plan/execution_plan_clone.c [hunk 4]
```diff
@@ -8,92 +8,69 @@
...
-	// Clone the op's children and add them the cloned children array.
-	for(uint i = 0; i < op->childCount; i++) {
-		OpBase *cloned_child = _ExecutionPlan_CloneOperations(orig_plan, clone_plan,
-															  op->children[i]);
-		// Assumption: all the children are either in the same segment or on a different segment.
-		if(!cloned_child) break;
-		ExecutionPlan_AddOp(clone, cloned_child);
+	clone->record_map = raxClone(template->record_map);
+	if(template->ast_segment) clone->ast_segment = AST_ShallowCopy(template->ast_segment);
+	if(template->query_graph) clone->query_graph = QueryGraph_Clone(template->query_graph);
```
Comment (swilly22) [inline src/execution_plan/execution_plan_clone.c:16]: I think we can avoid both `query_graph` and `connected_components` if we introduce ref count to the QueryGraph object.

File: src/execution_plan/execution_plan_clone.c (hunk 3 above)
Comment (jeffreylovitz) [reply src/execution_plan/execution_plan_clone.c:57]: This turns out to not be redundant, the `root` pointer is never touched by the cloning or the `New` op routines.

**Pull request merged**